server/etcdserver/api/v3rpc/watch.go

```diff
@@ -270,7 +270,7 @@ func (sws *serverWatchStream) recvLoop() error {
 			if !sws.isWatchPermitted(creq) {
 				wr := &pb.WatchResponse{
 					Header:       sws.newResponseHeader(sws.watchStream.Rev()),
-					WatchId:      creq.WatchId,
+					WatchId:      -1,
```

Comment:
Probably we should check the value of `creq.WatchId`, how about the logic something like below?
```
watchId := creq.WatchId
if watchId == int64(mvcc.AutoWatchID) {
	watchId = -1
}
```

Another choice is to make watchId starting with 1 instead of 0. But we need to fix some test cases, and it also change the default behavior, accordingly might have some impact on client applications which depend on the watch. 

I manually verified that this PR indeed can fix the issue https://github.com/etcd-io/etcd/issues/12385. 

But the overall design looks weird to me. The existing watch can continue to work even the simple token expired, because the client can retry the authentication. But creating new watch stream using exactly the same client doesn't work, it's a little strange. 

It seems that we need to get the overall design clarified firstly. WDYT @mitake

Reply:
I set  watchId starting with 1 at the beginning, all issues are resolved and do not conflict with `AutoWatchID` but test failed. If can fix test, I think it is better to start with 1.

I haven't found a place to set `creq.WatchId` yet, so it should always be mvcc.autowatchid?

Reply:
I think it's better to use -1 as an invalid watch ID for preserving the current behavior of watch API like what current PR does. How about introducing a new const like `AutoWatchID`?

Let me check the behavior introduced by this PR in a few days. I only tested a case which don't rely on it.

Reply:
> I haven't found a place to set creq.WatchId yet, so it should always be mvcc.autowatchid?

Yes, it's basically true. In theory, users do not necessarily use the etcd client SDK. When they construct the `wathRequest` themselves, then watchId might not be 0 anymore. But the possibility should be very low, and it isn't the recommended way. So probably we can assume that it's always `mvcc.AutoWatchID` for any `pb.WatchRequest_CreateRequest`. 

> I think it's better to use -1 as an invalid watch ID for preserving the current behavior of watch API like what current PR does.

Based on point above (watchId always be `mvcc.AutoWatchID` for any `pb.WatchRequest_CreateRequest`), it makes sense to always use `-1` as an invalid watch ID

Reply:
> In theory, users do not necessarily use the etcd client SDK. When they construct the wathRequest themselves, then watchId might not be 0 anymore. But the possibility should be very low, and it isn't the recommended way.

Yeah using non standard SDK is technically possible... but I think it's good to exclude it because it's quite irregular as you point out.

Reply:
Adding InvalidWatchID is a good idea to me. 

Regarding to adding a condition like` c.WatchID != InvalidWatchID`... it shouldn't happen otherwise it's a bug. So I think we should add a [Assert](https://github.com/etcd-io/etcd/blob/22cc6826064aabea365480e63c43b352c42eba07/client/pkg/verify/verify.go#L76).  For example [schedule.go#L92](https://github.com/etcd-io/etcd/blob/22cc6826064aabea365480e63c43b352c42eba07/pkg/schedule/schedule.go#L92)

Reply:
Yeah, I think using `Assert` is a better idea 👍

Reply:
I have revised it as suggested and any other suggestions for the Assert location?

Reply:
> I think `WatchResponse` with `Created && Canceled` and valid ID is possible in other paths, e.g. https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L308
> Could you update the assert condition for checking `CancelReason`? @kafuu-chino

The [Watch](https://github.com/etcd-io/etcd/blob/e9c53a0875aa048b79f1844d6cd09973691d1262/server/storage/mvcc/watcher.go#L108) always returns `-1` when there is an error. Please replace all the `-1` with the const `InvalidWatchID`. cc @kafuu-chino

Reply:
> I think `WatchResponse` with `Created && Canceled` and valid ID is possible in other paths, e.g. https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L308 Could you update the assert condition for checking `CancelReason`? @kafuu-chino

I think the `WatchId` shoud be `InvalidWatchID` if `err != nil`.

https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L310

For example: if we send duplicated `WatchId` from client, err will be `ErrWatcherDuplicateID`, the `WatchId` will cancel by server.

Then the `Assert` will not trigger unless there is a bug.
@mitake

Reply:
@ahrtr 
I tried to replace all the -1 I found , then squashed commit.

Reply:
@kafuu-chino I see, but actually I could trigger the assert in `TestKVCompact`. Let me check again. It's great if you can run the test on your side too.

Reply:
@mitake I run `TestKVCompact ` on my local system but no assert triggerred, did you update my later commit？

Reply:
@kafuu-chino yeah the assert is still failing like below:
```

goroutine 252 [chan receive]:
go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc.(*serverWatchStream).sendLoop.func1()
	/Users/mitake/go/src/go.etcd.io/etcd/server/etcdserver/api/v3rpc/watch.go:373 +0xac
panic({0x104c27660, 0x140036801f0})
	/usr/local/go/src/runtime/panic.go:838 +0x204
go.etcd.io/etcd/client/pkg/v3/verify.Assert(...)
	/Users/mitake/go/src/go.etcd.io/etcd/client/pkg/verify/verify.go:78
go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc.(*serverWatchStream).sendLoop(0x14000540f00)
	/Users/mitake/go/src/go.etcd.io/etcd/server/etcdserver/api/v3rpc/watch.go:473 +0xb10
go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc.(*watchServer).Watch.func1()
	/Users/mitake/go/src/go.etcd.io/etcd/server/etcdserver/api/v3rpc/watch.go:181 +0x28
created by go.etcd.io/etcd/server/v3/etcdserver/api/v3rpc.(*watchServer).Watch
	/Users/mitake/go/src/go.etcd.io/etcd/server/etcdserver/api/v3rpc/watch.go:180 +0x1e8

```

Anyway the assert isn't mandatory. Passing `InvalidWatchID` to delete() or looking it up from `ids` isn't harmful. I think it's ok to remove. Or we can update the condition to check `CancelReason` like `verify.Assert(c.CancelReason == "" && wid != mvcc.InvalidWatchID, "unexpected watchId: %d, wanted: %d, since CancelReason is empty", wid, mvcc.InvalidWatchID)`. How do you think?

Reply:
@mitake 
I only found two places to return both `Canceled` and `Created` are true and can trigger the assert.
1. https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L271
2. https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L308

What do you think which cause assert triggerred? Or i missed something and can you help to supplement?

I agree with :
> Passing `InvalidWatchID` to delete() or looking it up from ids isn't harmful. 

Because it will not affect the normal watch.

Does `c.CancelReason == "" ` equal to `c.Canceled == flase`? I think it means the watch will not be deleted and the `WatchId` will be a normal watch id. 

In my opinion, the purpose of this assert is that when a watch is newly created and an error is encountered ( both `Canceled` and `Created` are true), do not cause other normal watches to be deleted (`WatchId` MUST BE `InvalidWatchID`), it's hard to detect when it happens.

Reply:
@kafuu-chino In the case of `TestKVCompaction`, https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L308 is causing the assert. The condition is `c.Created == true && c.Canceled == false && c.WatchId == 0 && c.CancelReason == ""`.

On the second thought, the condition should be like `(c.Canceled || c.Created) && wid != mvcc.InvalidWatchID`? If it's false, the branches https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L468-L472 will be executed and unintentional operations against invalid watch ID will happen (although it's not harmful).

It's great if you can try to run the test case too: `TESTCASE=TestKVCompact TEST_OPTS="PASSES='integration'" make test`

Reply:
@mitake 
The assert condition really wrong and I've got the condition backwards.

If condition like `(c.Canceled || c.Created) && wid != mvcc.InvalidWatchID`, there will be problems here: https://github.com/etcd-io/etcd/blob/main/server/etcdserver/api/v3rpc/watch.go#L342

1. `c.Created == false` and `c.Canceled == false` 
2. then `(c.Canceled || c.Created) == false`
3. and `(c.Canceled || c.Created) && wid != mvcc.InvalidWatchID` == false
4. `condition == false` and assert will be trigger

How about `!(c.Canceled && c.Created) || wid == mvcc.InvalidWatchID`？

Reply:
@kafuu-chino Sorry the above my condition is very wrong... Especially if it includes `&& wid != mvcc.InvalidWatchID`, it cannot handle an invalid `wid` always. I guess I pasted a wrong version :(
I think your condition is correct. Sorry for the confusion.

Reply:
Let me review the entire PR, I think I can reply sometime tomorrow.